src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +340,138 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
+        {
+            progressTracker.AddItems(solution.ProjectIds.Count);
+            foreach (var projectId in solution.ProjectIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var project = solution.GetProject(projectId);
+                progressTracker.Description = project.Name;
+                var newProject = await FixProjectAsync(project, enabledFixIds, new ProgressTracker(), cancellationToken).ConfigureAwait(false);
+                solution = newProject.Solution;
+                progressTracker.ItemCompleted();
+            }
+
+            return solution;
+        }
+
+        private async Task<Project> FixProjectAsync(Project project, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (project.LanguageServices.GetService<ICodeCleanupService>() is null)
+            {
+                return project;
+            }
+
+            progressTracker.AddItems(project.DocumentIds.Count);
+            foreach (var documentId in project.DocumentIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var document = project.GetDocument(documentId);
+                progressTracker.Description = document.Name;
+                var fixedDocument = await FixDocumentAsync(document, enabledFixIds, new ProgressTracker(), cancellationToken).ConfigureAwait(false);
+                project = fixedDocument.Project;
+                progressTracker.ItemCompleted();
+            }
+
+            return project;
+        }
+
+        private async Task<Document> FixDocumentAsync(Document document, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            var codeCleanupService = document.GetLanguageService<ICodeCleanupService>();
+
+            var allDiagnostics = codeCleanupService.GetAllDiagnostics();
+
+            var enabedDiagnosticSets = ArrayBuilder<DiagnosticSet>.GetInstance();
+
+            foreach (var diagnostic in allDiagnostics.Diagnostics)
+            {
+                foreach (var diagnosticId in diagnostic.DiagnosticIds)
+                {
+                    if (enabledFixIds.IsFixIdEnabled(diagnosticId))
+                    {
+                        enabedDiagnosticSets.Add(diagnostic);
+                        break;
+                    }
+                }
+            }
+
+            var isRemoveUnusedUsingsEnabled = enabledFixIds.IsFixIdEnabled(RemoveUnusedImportsFixId);
+            var isSortUsingsEnabled = enabledFixIds.IsFixIdEnabled(SortImportsFixId);
+            var enabledDiagnostics = new EnabledDiagnosticOptions(enabedDiagnosticSets.ToImmutableArray(),
+                new OrganizeUsingsSet(isRemoveUnusedUsingsEnabled, isSortUsingsEnabled));
+
+            return await codeCleanupService.CleanupAsync(
```

Comment:
it it looks like. the very very first code clean will run on UI thread (or the thread this command is invoked)? and then subsequent code clean up will run in BG due to ConfigureAwait(false)? shouldn't we explicitly do Task.Run(...) in outter FixSolution,FixProjectAsync and etc to make sure inner FixSolutionAsync/FixProjectAsync/FixDocumentAsync runs from BG?

Reply:
hmm.. if that's true. I am not sure why you are doing switch main thread and then explicitly switching to thread pool up there? won't it be fine just letting it run things in whatever place it run?

Reply:
there is a theoradical difference I believe though since if things run on UI thread, even if user can cancel operation through wait dialog window, VS's UI will be freezed. if things run in BG while wait dialog is up, then VS UI is still responsive even if input is blocked.

but in practice, not sure whether that matter much.